Separate REPL evaluation from it's result; fixes #79196

dgozman opened the pull request:
REPL now has two elements: one for the evaluation text added immediately (EvaluationReplElement), and another for the evaluation result once available (Expression).

dgozman commented:
@isidorn Please take a look. I went with two separate objects strategy, as you were suggesting. Thank you!

isidorn commented:
@dgozman nice work, I really like it.
However I would prefer if we clean this up a bit. For example I feel like we are abusing the `Expression` class here since we only use the value. And we introduced an issue since the `toString` is now not correct in the Watch Expression view since the `Expression` is also used there.

Due to that I suggest the following:
1) Move all the REPL specific classes from `debugModel.ts` to `replModel.ts` (`SimpleReplElement`, `EvaluationReplElement`, `RawObjectReplElement`)
2) Instead of reusing the `Expression` class I suggest we introduce a new class which will live in `replModel.ts` and extend `ExpressionContainer` and will only be used in repl. It will not have the unneded stuff from `Expression`, and the `expression.evaluate` can be moved to a helper method that both of these can use
3) We get rid of stale css class names `output.expression` both in `repl.css` and `repl.ts` and we use our new names.

This new class and the `EvaluationReplElement` are pairs. Due to that we should give them nice name. Here are some ideas and I am open to suggestions:
* `ReplEvaluationInput` and `ReplEvaluationOutput`
* `ReplEvaluationExpression` and `ReplEvaluationResult`
* `ReplEvaluationInput` and `ReplEvaluationResult` (might be my favorite combination)

Let me know what you think. Thanks a lot!

dgozman commented:
@isidorn Thank you for detailed feedback! I followed your suggestions, please take another look.
- Went with `ReplEvaluationInput` and `ReplEvaluationResult`.
- Moved code to `replModel.ts`.
- Removed `.input.expression`, `.output.expression` and `.input-output-pair` from repl evaluations. We still need `.expression`, to pickup styling from `renderExpressionValue`. Also, there are still `.output.expression` left in other repl elements, which I'd like to not touch here. Is that ok?

isidorn commented on src/vs/workbench/contrib/debug/common/replModel.ts:
```diff
@@ -6,15 +6,107 @@
+	}
+}
+
+export class ReplEvaluationResult extends ExpressionContainer implements IReplElement, IExpression {
```
I do not think `ReplEvaluationResult` should implement `IExpression` and it should not have a `public name` which we do not use anywhere.
I see that you are passing `ReplEvaluationResult` to `renderExpressionValue`, but can we pass just it's `value` instead.
I think we get rid of the `name` it will get clearer that the ReplEvaluationResult is just the result

I might be missing something here, but idealy we can simplify this.

isidorn commented on src/vs/workbench/contrib/debug/common/replModel.ts:
```diff
@@ -24,10 +116,14 @@ export class ReplModel {
+		const result = new ReplEvaluationResult(name);
+		const evaluationResult = await ExpressionContainer.evaluate(name, result, this.session, stackFrame, 'repl');
+		if (evaluationResult) {
+			result.type = evaluationResult.type || result.type;
```
Is it possible to move the `type` up to `ExpressionContainer` so we do not have this code duplicated and just put it in the static helper function?

isidorn commented on src/vs/workbench/contrib/debug/browser/repl.ts:
```diff
@@ -835,10 +859,10 @@ class ReplAccessibilityProvider implements IAccessibilityProvider<IReplElement>
 			return nls.localize('replVariableAriaLabel', "Variable {0} has value {1}, read eval print loop, debug", element.name, element.value);
 		}
-		if (element instanceof Expression) {
+		if (element instanceof ReplEvaluationResult) {
```
As mentione in the other comment `ReplEvaluationResult` should not have a name, since the name is now represented by the `ReplEvaluationInput`.
Thus only the value should be read out here, so data is not duplicated to the user using a screen reader.

isidorn commented:
This looks great, thanks a lot!
I have left some comments inline in the code. Can you please tackle those and then we can merge this in?  Apart from that:
* Can you please merge with latest from master and resolve conflicts? There is one simple conflict in `debugSession.ts`
* You did changes in the `getHeight` method and they make sense, however can you please test that the REPL scroll position behaves as expected when new output comes. Here's an example of a bug https://github.com/microsoft/vscode/issues/71737 which we can potentially break by changing the `getHeight`
* Did you run the smoke test so we verify we did not break anything with the css names? Here's how to run it https://github.com/Microsoft/vscode/blob/master/test/smoke/README.md
* Is it possible to write a unit test that captures these new behavior so we are sure we do not break it in the future? Adding an element immediatly adds it to the repl, and once it is evaluated we check there is an additional element?
* Not touching `.output.expression` makes sense, that is not directly related to this PR.

Thanks again!

isidorn commented:
Great work, I like that we moved those attributes up to IExpressionContainer.
As for the unit test to test the whole debug session flow - we do not have. Since that would require a real connection to the adapter, and spawing processes which felt liek out of scope for unit tests. Though we should probably invest in writing some integration tests for debugging.

I also tried out your changes and tehy look awesome to me. Merging in. Thanks!
